controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -18,30 +20,47 @@
 import io.pravega.controller.store.stream.tables.Data;
 import io.pravega.controller.store.stream.tables.HistoryRecord;
 import io.pravega.controller.store.stream.tables.IndexRecord;
-import io.pravega.controller.store.stream.tables.SegmentRecord;
 import io.pravega.controller.store.stream.tables.State;
 import io.pravega.controller.store.stream.tables.TableHelper;
-import io.pravega.client.stream.StreamConfiguration;
 import lombok.extern.slf4j.Slf4j;
 import lombok.val;
+import org.apache.commons.lang.SerializationUtils;
 import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
 
 import java.util.AbstractMap;
 import java.util.AbstractMap.SimpleImmutableEntry;
 import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionException;
 import java.util.concurrent.CompletionStage;
+import java.util.function.Supplier;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
 @Slf4j
 public abstract class PersistentStreamBase<T> implements Stream {
 
+    private static final Map<State, List<State>> STATE_TRANSITIONS;
+
+    static {
+        Map<State, List<State>> map = new HashMap<>();
```

Comment:
Unmodifiable map is just a wrapper that prevents modification. It does not provide any synchronization. Without a memory barrier like a lock one thread cannot assume that it can see the full work of another thread. For example, it is considered fair for the compiler to re-order line 62 up to line 54 and perform that assignment first, and another thread could see a partially populated map, or worse one being populated. This can result in very strange and bad behaviour: https://mailinator.blogspot.com/2009/06/beautiful-race-condition.html